bin/spark-class

```diff
@@ -65,24 +65,25 @@ fi
 # characters that would be otherwise interpreted by the shell. Read that in a while loop, populating
 # an array that will be used to exec the final command.
 #
-# The exit code of the launcher is appended to the output, so the parent shell removes it from the
-# command array and checks the value to see if the launcher succeeded.
-build_command() {
-  "$RUNNER" -Xmx128m -cp "$LAUNCH_CLASSPATH" org.apache.spark.launcher.Main "$@"
-  printf "%d\0" $?
-}
+# To keep both the output and the exit code of the launcher, the output is first converted to a hex
+# dump which prevents the bash from getting rid of the NULL character, and the exit code retrieved
+# from the bash array ${PIPESTATUS[@]}.
+#
+# Note that the seperator NULL character can not be replace with space or '\n' so that the command
+# won't fail if some path of the user contain special characher such as '\n' or space
+#
+# Also note that when the launcher fails, it might not output something ending with '\0' [SPARK-16586]
+_CMD=$("$RUNNER" -Xmx128m -cp "$LAUNCH_CLASSPATH" org.apache.spark.launcher.Main "$@"|xxd -p|tr -d '\n';exit ${PIPESTATUS[0]})
```

Comment:
If you further add a line like `echo ${CMD[@]}`  after `if [ "$LAUNCHER_EXIT_CODE" != 0 ]; then`, it will output something like `There is an error1`, which will report the error happened but the ending error code `1` looks ugly.